src/detect-engine-build.c

```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
+        /* Treat right boundary as single point to avoid creating unneeded
+         * ranges later on.
+         * TODO merge the extra entry created in the rulegroup
+         * for such points in *CreateGroupedPortList* fn */
+        unique_list[p->port2] = SINGLE_PORT;
+        size_list++;
+    }
+    return size_list;
+}
+
+/**
+ * \brief Function to set the *final* unique port points and save them
+ *        for later use. The points are already sorted because of the way
+ *        they have been retrieved and saved earlier for use at this point.
+ *
+ * \param unique_list List of the unique port points to be used
+ * \param size_unique_arr Number of unique port points
+ * \param final_arr List of the final unique port points to be created
+ */
+static inline void SetFinalUniquePortPoints(
+        const uint8_t *unique_list, const uint16_t size_unique_arr, UniquePortPoint *final_arr)
+{
+    for (int i = 0, j = 0; i < MAX_PORT_VAL; i++) {
+        DEBUG_VALIDATE_BUG_ON(j > size_unique_arr);
+        if (unique_list[i] == RANGE_PORT) {
+            final_arr[j].port = (uint16_t)i;
+            final_arr[j++].single = false;
+        } else if (unique_list[i] == SINGLE_PORT) {
+            final_arr[j].port = (uint16_t)i;
+            final_arr[j++].single = true;
+        }
+    }
+}
+
+/**
+ * \brief Function to create the list of ports with the smallest ranges
+ *        by resolving overlaps and end point conditions. These contain the
+ *        correct SGHs as well after going over the interval tree to find
+ *        any range overlaps.
+ *
+ * \param de_ctx Detection Engine Context
+ * \param unique_list Final list of unique port points
+ * \param size_list Size of the unique_list
+ * \param it Pointer to the interval tree
+ * \param list Pointer to the list where final ports will be stored
+ *
+ * \return 0 on success, -1 otherwise
+ */
+static inline int CreatePortList(DetectEngineCtx *de_ctx, const uint8_t *unique_list,
+        const uint16_t size_list, SCIntervalTree *it, DetectPort **list)
+{
+    /* Only do the operations if there is at least one unique port */
+    if (size_list > 0) {
+        uint16_t port, port2;
+        UniquePortPoint *final_unique_points =
+                (UniquePortPoint *)SCCalloc(size_list, sizeof(UniquePortPoint));
+        if (final_unique_points == NULL)
+            return -1;
+        SetFinalUniquePortPoints(unique_list, size_list, final_unique_points);
+        /* Handle edge case when there is just one unique port */
+        if (size_list == 1) {
+            port = port2 = final_unique_points[0].port;
+            PISearchOverlappingPortRanges(de_ctx, port, port2, &it->tree, list);
+        } else {
+            UniquePortPoint *p1 = NULL, *p2 = NULL;
+            p1 = &final_unique_points[0];
+            p2 = &final_unique_points[1];
+            port = p1 ? p1->port : 0; // just for cppcheck
+            port2 = p2->port;
+            for (uint16_t i = 1; i < size_list;) {
```

Comment:
what is the max value of `size_list`?

Reply:
65536 if we have all the ports possible available in our ruleset

Reply:
size_list can't represent that as it has u16. Probably best to switch to int or uint32_t

Reply:
That was stupid. Sorry. Half the places, I have taken care of the limit and half are uncaught. Thank you!